packages/lib-taquito/index.ts

```diff
@@ -0,0 +1,111 @@
+import { DfnsApiClient, DfnsError } from '@dfns/sdk'
+import { Prefix, b58cencode, buf2hex, getPkhfromPk, hex2buf, mergebuf, prefix } from '@taquito/utils';
+import { Signer } from '@taquito/taquito'
+import { GetWalletResponse, GenerateSignatureResponse } from '@dfns/sdk/types/wallets'
+import { TxWatermark } from './constant'
+import { error } from 'console';
+
+export type DfnsWalletOptions = {
+  walletId: string
+  dfnsClient: DfnsApiClient
+  maxRetries?: number
+  retryInterval?: number
+}
+
+type WalletMetadata = GetWalletResponse
+
+const assertSigned = (res: GenerateSignatureResponse) => {
+  if (res.status === 'Failed') {
+    throw new DfnsError(-1, 'signing failed', res)
+  } else if (res.status !== 'Signed') {
+    throw new DfnsError(
+      -1,
+      'cannot complete signing synchronously because this wallet action requires policy approval',
+      res
+    )
+  }
+}
+
+export class DfnsWallet implements Signer{
+  private options: Required<DfnsWalletOptions>
+  private readonly dfnsClient
+
+  private constructor(private metadata: WalletMetadata, options: DfnsWalletOptions, private tezosPubKey: string) {
+    this.dfnsClient = options.dfnsClient
+  }
+
+  public static async init(options: DfnsWalletOptions) {
+    const { walletId, dfnsClient } = options
+    const res = await dfnsClient.wallets.getWallet({ walletId })
+
+    if (res.status !== 'Active') {
+      throw new DfnsError(-1, 'wallet not active', { walletId, status: res.status })
+    }
+
+    if (res.network !== 'Tezos' && res.network !== 'TezosGhostnet') {
+      throw new DfnsError(-1, 'wallet is not bound to Tezos or TezosGhostnet', { walletId, network: res.network })
+    }
+
+    // get the public key encoded with the right tezos prefix.
+    const tezosPublicKey = this.getEncodedPublicKey(res.signingKey.publicKey, res.signingKey.scheme)
+
+    return new DfnsWallet(res, options, tezosPublicKey)
+  }
+
+  private static getEncodedPublicKey(publicKey: string, scheme: string): string {
+    switch (scheme) {
+      case 'EdDSA':
+        return b58cencode(publicKey, prefix[Prefix.EDPK])
+      case 'ECDSA':
+        return b58cencode(publicKey, prefix[Prefix.SPPK])
+      default:
+        throw new Error(`KeyScheme.${scheme} is not supported for tezos`)
+    }
+  }
+
+  public get address(): string {
+    // Tezos-bound wallets will have a tezos address
+    return this.metadata.address!
+  }
+
+  // publicKey implements the Taquito Signer interface
+  publicKey(): Promise<string> {
+    return Promise.resolve(this.tezosPubKey)
+  }
+
+  // publicKeyHash implements the Taquito Signer interface
+  publicKeyHash(): Promise<string> {
+    return Promise.resolve(this.metadata.address!)
+  }
+
+  // secretKey implements the Taquito Signer interface
+  secretKey(): Promise<string | undefined> {
+    throw undefined;
+  }
+
+  // sign implements the Taquito Signer interface
+  // For now, we only want to support transaction signing
```

Comment:
correct me if I'm misremembering, but I think we can support all signing. the logic would be

1. if magic byte is tx watermark, call sign with `kind: Transaction`.
2. otherwise, we can prepend magic byte, blake2b hash, and sign with `kind: Hash`.

maybe in the future if there's demand, we can support more kinds, but for now, this should work right? no need to throw error.